cephfs-mirror: incremental sync

vshankar opened the pull request:
Requires PR #40810 to build and use.

Snap data is incrementally transferred to remote file system based on mtime/ctime differences. This does away with cleaning up remote file system data on each snap sync. Whenever possible, changes are identified by scanning local snapshot pair. On the other hand, when last synced ("previous") snapshot cannot be used for comparison, changes are identified by comparing "current" snapshot data to remote filesystem directory root (for the directory in question).

Also, this relies to *at() libcephfs APIs to ensure correctness of the snapshot being synchronized FTW.

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

vshankar commented:
Build failures are due to requirement of PR #40810.

batrick commented:
````
./tasks/cephfs/test_mirroring.py:907:19: F541 f-string is missing placeholders
./tasks/cephfs/test_mirroring.py:947:13: F841 local variable 'p' is assigned to but never used
````

mchangir commented on src/tools/cephfs_mirror/PeerReplayer.h:
```diff
@@ -45,6 +45,30 @@ class PeerReplayer {
   inline static const std::string SERVICE_DAEMON_FAILED_DIR_COUNT_KEY = "failure_count";
   inline static const std::string SERVICE_DAEMON_RECOVERED_DIR_COUNT_KEY = "recovery_count";
 
+  typedef std::pair<std::string, uint64_t> Snapshot;
+
```
nit: you might want to use the `using` directive here instead of `typedef`

vshankar replied:
done

mchangir reviewed (commented):
just a nit

vshankar commented:
https://pulpito.ceph.com/vshankar-2021-04-30_17:19:54-fs-wip-cephfs-mirror-incremental-sync-20210430-093812-distro-basic-gibba/

Failure is: https://tracker.ceph.com/issues/50224

batrick commented:
... and rebased again

vshankar commented:
rebase and updated -- also included changes to sync file mode.

mchangir reviewed (commented):
small request

vshankar commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -560,29 +574,26 @@ int PeerReplayer::remote_mkdir(const std::string &local_path,
-                              const struct ceph_statx &stx) {
-  dout(10) << ": dir_path=" << dir_path << ", local_path=" << local_path
-           << ", remote_path=" << remote_path << dendl;
+int PeerReplayer::remote_copy(const std::string &dir_root,  const std::string &epath,
+                              const struct ceph_statx &stx, const FHandles &fh) {
```
copy_to_remote() seems apt. thx!

vshankar commented:
rebased and updated.

test run: https://pulpito.ceph.com/vshankar-2021-05-26_04:31:49-fs-wip-cephfs-mirror-incremental-sync-20210525-084706-distro-basic-gibba/

@mchangir @fullerdj

mchangir commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -975,7 +975,7 @@ int PeerReplayer::synchronize(const std::string &dir_path, uint64_t snap_id,
   auto snap_id_str{stringify(snap_id)};
   snap_metadata snap_meta[] = {{PRIMARY_SNAP_ID_KEY.c_str(), snap_id_str.c_str()}};
   r = ceph_mksnap(m_remote_mount, dir_path.c_str(), snap_name.c_str(), 0755,
```
`snap_metadata` could have been named with the suffix `_t` as `snap_metadata_t`

vshankar replied:
hmm ok. Let's fix this in a separate pr?

mchangir reviewed (commented):
nit

mchangir reviewed (approved):
LGTM
